core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -1882,32 +1882,31 @@ public FileInfo createFile(AlluxioURI path, CreateFileContext context)
    * @param inodePath the path to be created
    * @param context the method context
    */
-  private void overwriteFileInternal(RpcContext rpcContext, LockedInodePath inodePath,
-                                     CreateFileContext context)
+  private void deleteFileIfOverwrite(RpcContext rpcContext, LockedInodePath inodePath,
+      CreateFileContext context)
       throws FileDoesNotExistException, IOException, InvalidPathException,
       FileAlreadyExistsException {
     if (inodePath.fullPathExists()) {
       Inode currentInode = inodePath.getInode();
-      if (!context.getOptions().getOverwrite()) {
+      if (!context.getOptions().hasOverwrite() || !context.getOptions().getOverwrite()) {
         throw new FileAlreadyExistsException(
-            "Not allowed to create() (overwrite=false) for existing Alluxio path: "
-                + inodePath.getUri());
+            ExceptionMessage.CANNOT_OVERWRITE_FILE_WITHOUT_OVERWRITE.getMessage(
+                inodePath.getUri()));
       }
       // if the fullpath is a file and the option is to overwrite, delete it
-      if (currentInode instanceof InodeFile) {
+      if (currentInode.isDirectory()) {
+        throw new FileAlreadyExistsException(
+            ExceptionMessage.CANNOT_OVERWRITE_DIRECTORY.getMessage(inodePath.getUri()));
+      } else {
         try {
           deleteInternal(rpcContext, inodePath, DeleteContext.mergeFrom(
               DeletePOptions.newBuilder().setRecursive(true)
                   .setAlluxioOnly(!context.isPersisted())), true);
           inodePath.removeLastInode();
         } catch (DirectoryNotEmptyException e) {
-          // IGNORE, this will never happen
```

Comment:
nit: add a comment that we should not reach here like
`// Should not reach here`

Reply:
done